### ISSUE TITLE
Section variables are in scope of operator arguments

### ISSUE THREAD

**Issue by ncfavier** [agda/agda#7041]
```agda
open import Agda.Builtin.Equality

id : {A : Set} → A → A
id x = x

infixr 9 _∘_
_∘_ : {A B C : Set} → (B → C) → (A → B) → A → C
(f ∘ g) x = f (g x)

cong : {A B : Set} {x y : A} → (f : A → B) → x ≡ y → f x ≡ f y
cong f refl = refl

test : {A : Set} {f g : A → A}
     → f ∘ f ≡ id → f ∘ f ∘ g ≡ g
test p = cong (_∘ ?) p
```

results in the following error:

```
Failed to solve the following constraints:
  ?0 (p = p) (section = id) x = g x : A (blocked on _45)
  ?0 (p = p) (section = (f ∘ f)) x = g x : A (blocked on _45)
```

where `section` is the variable in the desugared form `cong (λ section → section ∘ ?)`.

The constraints could be solved if `section` weren't in scope, which it has no reason to be in my opinion since there's no way I can refer to it.

**Feature request**: somehow prune the `section` variable away from metavariables appearing inside operator sections.

**Comment by sattlerc** [comment]
For the purpose of unification, it seems better to desugar
```
<operator section with non-placeholders>
```
to
```
(λ <non-placeholder arguments> <placeholder arguments> → <fully applied operator>) <non-placeholder arguments>
```
That would solve this problem.

**Comment by sattlerc** [comment]
The suggestion from the previous post doesn't work in the following example:
```agda
postulate
  A : Set
  _,_,_ : (X : A → Set) (a : A) (x : X a) → Set
  B : Set
  b : B

-- Works.
x = (λ a → B) ,_, b

-- This desugaring doesn't type-check.
y = (λ X x a → _,_,_ X a x) (λ a → B) b
```

**Comment by sattlerc** [comment]
The proper solution is to use let-bindings instead of β-redexes.

Consider a function F of arity I (finite total order). Consider a section (application with potential placeholders) F((a_i)_{i∈I}) where a_i is either a hole or a term. We desugar it to:
```
let (x_i = a_i)_{i∈I, a_i non-hole} in λ(x_i)_{i∈I, a_i hole}. F(x)
```
or using let-bindings in lambda-abstractions:
```
λ (let (x_i = a_i)_{i∈I, a_i non-hole}) (x_i)_{i∈I, a_i hole}. F(x)
```


Example from above:
```agda
postulate
  A : Set
  _,_,_ : (X : A → Set) (a : A) (x : X a) → Set
  B : Set
  b : B

-- Current situation: works, but is not transparent with respect to unification.
x = (λ a → B) ,_, b

-- This desugaring doesn't type-check.
-- y = (λ X x a → _,_,_ X a x) (λ a → B) b

-- This desugaring works and is transparent with respect to unification.
z = λ (let X = λ a → B; x = b) a → _,_,_ X a x
```

**Comment by sattlerc** [comment]
Here is a smaller example of this bug.

```agda
postulate
  A B : Set
  _&_ : A → A → B

f : A → B
f = _& {!!}
```

The hole shows:
```
Goal: A
————————————————————————————————————————————————————————————
section : A
```

Moreover, C-c C-a will happily fill it with `section`, but refuse to reload.

**Comment by sattlerc** [comment]
Better example:

```agda
postulate
  A : Set

data _≡_ : (x y : A) → Set where
  refl : (a : A) → a ≡ a

x : let P = ? ≡_ in (a : A) → P a
x = refl
```

Agda will solve the hole with `section`, which is not in scope.

This can be turned into a failure test case:

```agda
postulate
  A : Set

data _≡_ : (x y : A) → Set where
  refl : (a : A) → a ≡ a

-- This should fail.
y : let P = _ ≡_ in (a : A) → P a
y = refl
```

**Comment by andreasabel** [comment]
@ncfavier wrote:
> ```agda
> test p = cong (_∘ ?) p
> ```

Not 100% sure about _bug_, maybe this is more in the _surprise_ category.

It is completely sound if you consider `(_∘ ?)` as sugar for `\ section -> section ∘ ?`.
Maybe not very pretty...

Desugaring before type-checking is one of the great evils which many decorated PL researchers and teachers have fallen prey to.  (I am tempted to mention SML-NJ...)

**Comment by sattlerc** [comment]
This is a scoping bug (even before type-checking).

The precise desugaring of `a •_` is `λ b → a[weaken] • b` where `weaken` is the weakening substitution lifting `a` to the context including `b`. That is, `a` is not in scope of the variable `b`, but `a[weaken]` is.

Unfortunately, Agda conflates `a` and `a[weaken]`. The variable `b` is chosen fresh before metavariables are filled in `a`, so it is not fresh for the metavariable solutions.

If we emulate the weakening using a let-binding as I wrote above, then I don't know how to prevent leaking the let-variable to the user. Is there some category of let-bindings in the abstract syntax that get unfolded when converted to concrete syntax?

**Comment by mietek** [comment]
Bumping as I just ran into this as well; thanks @ncfavier.
